src/cc/BPF.cc

```diff
@@ -0,0 +1,398 @@
+/*
+ * Copyright (c) 2016 Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cstdio>
+#include <exception>
+#include <linux/bpf.h>
+#include <memory>
+#include <unistd.h>
+#include <vector>
+
+#include "bcc_syms.h"
+#include "bpf_module.h"
+#include "libbpf.h"
+#include "perf_reader.h"
+
+#include "BPF.h"
+
+namespace ebpf {
+
+  std::string uint_to_hex(uint64_t value) {
+    std::stringstream ss;
+    ss << std::hex << value;
+    return ss.str();
+  }
+
+  std::string sanitize_str(
+    std::string str, bool (*validator)(char), char replacement = '_'
+  ) {
+    for (size_t i = 0; i < str.length(); i++)
+      if (!validator(str[i]))
+        str[i] = replacement;
+    return str;
+  }
+
+  BPF::BPF(
+    const std::string& bpf_program,
+    unsigned int debug_flag, const char* cflags[], int ncflags
+  ) :
+    bpf_module_(new BPFModule(debug_flag))
+  {
+    if (bpf_module_->load_string(bpf_program, cflags, ncflags) != 0)
+      throw std::runtime_error("Unable to initialize BPF program");
+  };
+
+  BPF::~BPF() {
+    for (auto it : kprobes_)
+      if (!detach_kprobe_event(it.first, it.second))
+        fprintf(
+          stderr, "Failed to detach kprobe event %s in BPF destructor\n",
+          it.first.c_str()
+        );
+
+    for (auto it : uprobes_)
+      if (!detach_uprobe_event(it.first, it.second))
+        fprintf(
+          stderr, "Failed to detach uprobe event %s in BPF destructor\n",
+          it.first.c_str()
+        );
+    for (auto it : tracepoints_)
+      if (!detach_tracepoint_event(it.first, it.second))
+        fprintf(
+          stderr, "Failed to detach Tracepoint %s in BPF destructor\n",
+          it.first.c_str()
+        );
+  }
+
+  bool BPF::attach_kprobe(
+    const std::string& kernel_func,
+    const std::string& probe_func,
+    bpf_attach_type attach_type,
+    pid_t pid, int cpu, int group_fd,
+    perf_reader_cb cb, void* cb_cookie
+  ) {
+    int probe_fd = load_func(probe_func, BPF_PROG_TYPE_KPROBE);
+    if (probe_fd == -1) {
+      fprintf(stderr, "Unable to load probe function %s\n", probe_func.c_str());
```

Comment:
you have a strange mix of stderr prints and exceptions... IMO exceptions are the way to go although i see i might be alone in thinking that